### ISSUE TITLE
Ella -> core + plugins

### ISSUE THREAD

**Issue by xaralis** [ella/ella#55]
We need to decide on conventions for ella plugins and which modules will be kept in core after the process. Seems like the only module with direct ties to core should be **photos** (where I would highly appreciate to use some interface to allow for [sorl-thumbnail](https://github.com/sorl/sorl-thumbnail) as I stated before).

Moreover, it is needed to decide on who will do the thing. In Fragaria, we would like to handle polls, since we use them quite often. Major refactoring will be most likely included even though we originally proposed to postpone it.

@whit said, he would like to keep following naming convention for all ella plugins: `ella_[PLUGIN_NAME]`.

So there would be `ella_articles`, `ella_polls` etc.

This leads to some minor troubles with DB table names and references in the code itself. There will have to be migrations to rename all the tables.

Please, share your opinions here. We would like to start working on this next week.

**Comment by xaralis** [comment]
See #56 too.

**Comment by honzakral** [comment]
Why sorl-thumbnail? it doesn't do anything ella's formatter doesn't do. I am not against it, any added flexibility is nice, but I am yet to see a valid reson for this.

I am for maintaining articles in core, they are very small and cms without articles make no sense at all. (InfoBoxes however should go).

Also I am absolutely fine with just the packages being called ella_\* and the app names remaining without change - there is absolutely no need to have the word absoluteley everywhere.

**Comment by whit** [comment]
sorl-thumbnail is much cleaner and easier to use for target group - coders/designers. My experience with Ella's formats is that to maintain more than single site (cleanly) is practically impossible. There are much duplicities, one must be care of format naming, will have formats that differs in only once boolean parameter etc.

With sorl, _designer_ have good, descriptive snippet in code and may not be interested in anything else (db, admin...).

I propose to keep only few named _format shortcuts_ for easier redaction (big/small square/rectangle etc.) on top of sorl (or similar app).

Also I think that ella.photos should be replaceable or refactored with respect to basic categorization and multiple sites.

**Comment by xaralis** [comment]
I agree, almost completely with @whit. The main issue here is that we basically don't need Ella's formater. There is no need to maintain our solution when there is a superior one. The switch is nothing painful and sorl provides so so many more features than Ella does (like exceptionally great dummy images for the means of development)... 

Let's do it the way @whit proposed, create some simple wrapper over the sorl.

**Comment by xaralis** [comment]
I agree on keeping articles in core. @HonzaKral is right that CMS without articles is hard to call a CMS...

**Comment by honzakral** [comment]
Well, you can use sorl easily without using formatedPhot - just use the Photo model and sorl-thumb. Just please leave the formatting stuff in there, until sorl can figure out how to easily do important_box, manually override crop coordinates for individual formats or upload entirely different files for any format we need the current functionality to remain.

btw for dummy images, just use http://lorempixel.com/ :)

To sumarize: I am OK with the possibility to use sorl on top of the Photo model, but I want to keep Formats and FormatedPhoto with current features. If you provide another API for solr or a switch to use sorl as a backend thats great, but don't take my features away please.

**Comment by xaralis** [comment]
We personally found out, that most of stuff current formatter allows is hardly ever required as writers tend to force us to turn of all cropping completely beacuse they are unable to have full control over the photo look. We can keep it, but were mostly likely aren't gonna use it anyway.

For dummy images, have look [here](http://thumbnail.sorl.net/reference/settings.html#thumbnail-dummy), it's quite a big improvement for development process :)

**Comment by martinbilek** [comment]
**Article**
I'm for maintaining articles separately from core. Is there any reason to be part of core? If not, then let's make project transparent and keep core as small as possible. I can imagine article-less Ella project using only photo-galleries.

**Naming convention**
I prefer naming convention like `ella_[PLUGIN_NAME]`, so it's transparent for developers. Also module names will be more unique with `ella_` prefix so there are no conflicts with external apps.

**Comment by honzakral** [comment]
@martinbilek  Articles are one model and that's it - to maintain that as a separate project seems like a huge overkill. Also I like the idea that we have a bare-minimum example of a Publishable object in ella proper.

That being said, do with it as you please, I can always create my own app containing one model and if articles are no longer in ella, I can create it exactly as I would want it to look :)

**Comment by xaralis** [comment]
If we keep it separate, it would be cleaner in principle of divided core and plug-in system. But still, it feels more logical to keep it in core since this app is used in 99% of Ella projects (I have actually never seen a one not using it).

This would make only these parts of _core_:
- real core itself
- photos
- utils - which need a major cleanup, lot of mess in there
- articles

As for naming conventions, I vote for renaming apps not only as packages (we would have to get rid of ella. prefix anyway), but also _in the code itself_. It's just cleaner and more readable for users not mentioning it's more conflict-prone as @martinbilek said. We can keep the table names even though I would be happier if we provided migrations to alter these too.

**Comment by xaralis** [comment]
I also vote for keeping **positions** in core as it's used in almost every Ella project.

**Comment by martinbilek** [comment]
If article will be part of core, than positions also make sense to be part too. There is planned meeting on Thursday with @whit, so we will make a decision.

**Comment by honzakral** [comment]
yep, core, photos, articles and positions seem good choices to me.

**Comment by xaralis** [comment]
Imports now live in https://github.com/ella/ella-imports.

Galleries now live in https://github.com/ella/ella-galleries.

**Comment by xaralis** [comment]
Comments now live in https://github.com/ella/ella-comments.

**Comment by xaralis** [comment]
Polls now live in https://github.com/ella/ella-polls.

**Comment by xaralis** [comment]
What to do with ellatagging?

There is a dependency in `RelatedManager` folded in try-except... We have discussed this with @HonzaKral yesterday and agreed that the best solution would be to provide some hooks for `RelatedManager` that each Ella plugin could use. Something like this:

``` python
# settings.py

RELATED_HOOKS = (
    'ella_tagging.hooks.related_hook',
    '...'
)

# in ella_tagging.hooks
def related_hook(base_qset):
    # do something with base_qset
    return base_qset
```

The hooks should accept `QuerySet` and return modified `QuerySet` too. This would be hard to achieve with current tagging app though (see #33).

**Comment by xaralis** [comment]
Related were finally solved by using following changes:

### Related finders

We use so called `RELATED_FINDERS`. These are functions, that know how to find objects related to another object and have following signature:

``` python
def related_finder(obj, count, mods=[], only_from_same_site=True):
    ...
```

Related finders are defined in application settings using `RELATED_FINDERS` and `DEFAULT_RELATED_FINDER` attributes. Defaults are set like this:

``` python
# in ella/core/conf.py
RELATED_FINDERS = {
    'default': 'ella.core.related_finders.related_and_category',
    'directly': 'ella.core.related_finders.related_only'
}
DEFAULT_RELATED_FINDER = 'default'
```

`RELATED_FINDERS` are be named. More on this later.

### Related manager

`RelatedManager`'s `get_related_for_object` method has been updated to accept `finder` parameter which defaults to `DEFAULT_RELATED_FINDER`. Instead of doing all work by itself, it proxies the logic to the chosen `finder`.

### Related template tag

`related` template tag has been updated to allow for specifing named `finder` when required. This enables a lot of custom querying super simple way, e.g.: by author, by source... Everything that comes to mind by simply defining a custom function :) Cute - isn't it?

Example usage:

```
{% related 10 directly articles.article, galleries.gallery for object as related_list %}
```

### Ella plugin integration

Plugins can easily create their own finders and therefore, it is super simple to replace/update related logic. Only required action is to list the used `finder` in application settings as stated above. This will be used in `ella_tagging` project so that `with_tagging` finder that also uses tags is gonna be defined.

### Utility for creating custom related

I have also added function `fillup_from_category` which accepts **related list** and all parameters in the same way as normal finders do and fills up the incoming **related list** with objects listed in same category if there is not enought items to fullfill prescribed **count** parameter. This simplifies the creation of custom `finders`.

**Comment by xaralis** [comment]
Please, share your thoughts and see 6fc3180.

**Comment by xaralis** [comment]
Tagging now lives in https://github.com/ella/ella-tagging.

This was the last module to leave the core. Core is nice and slim project now. Looks like we are done here!

**Comment by honzakral** [comment]
I would like for the related finders to be a list so if I want 10 related and the first function only gives me 5, only then will the second function be run.

Then I can integrated with tagging, search (see more_like_this functionality in haystack) etc.

great work everyone, I like our new Ella :)

**Comment by xaralis** [comment]
@HonzaKral that sounds even better, I'll update the solution to behave like you described.

**Comment by xaralis** [comment]
Done. Now, related finders are defined like this:

``` python
RELATED_FINDERS = {
    'default': (
        'ella.core.related_finders.directly_related',
        'ella_tagging.related_finders.related_by_tags',
        'ella.core.related_finders.related_by_category',
    )
}

```

Each function has signature:

``` python
def related_finder(obj, count, collected_so_far, mods=[], only_from_same_site=True):
    ....
```

Where `collected_so_far` is list containing objects already collected during the process. Each function is now responsible only for it's part, e.g. tagging app only provides search by tags and doesn't care about the rest.

This whole stuff now enables simple custom fallback and great flexibility!

**Comment by xaralis** [comment]
Closing, solved.
